File: backend/app/api/v1/videos.py
from fastapi import APIRouter, File, Form, HTTPException, UploadFile, status

from app.core.deps import AdminUser
from app.schemas.item import ReassignRequest, RotateRequest, VideoSplitIn
from app.services import image_import as image_import_service
from app.services import import_coco as import_coco_service
from app.services import item as item_service
from app.services import project as project_service
from app.services import user as user_service
from app.services import video as video_service

router = APIRouter()


def _require_project(project_id: int):
    project = project_service.get(project_id)
    if not project:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Project not found")
    return project


def _require_user(user_id: int):
    user = user_service.get_by_id(user_id)
    if not user:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Assignee does not exist")
    return user


@router.post(
    "/projects/{project_id}/videos",
    status_code=status.HTTP_201_CREATED,
    tags=["videos"],
)
async def upload_video(
    project_id: int,
    current_user: AdminUser,
    file: UploadFile = File(...),
    fps: float = Form(...),
    assignee_id: int | None = Form(None),
    rotation: int = Form(0),
    resize_mode: str = Form("pad"),
) -> dict:
    _require_project(project_id)
    if assignee_id is not None:
        _require_user(assignee_id)
    try:
        return video_service.extract_frames(
            project_id,
            file.file,
            file.filename or "video",
            fps,
            rotation=rotation,
            assignee_id=assignee_id,
            resize_mode=resize_mode,
        )
    except ValueError as e:
        code = (
            status.HTTP_413_REQUEST_ENTITY_TOO_LARGE
            if "larger than" in str(e).lower()
            else status.HTTP_400_BAD_REQUEST
        )
        raise HTTPException(code, str(e))
    except RuntimeError as e:
        raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, str(e))


@router.post(
    "/projects/{project_id}/import-coco",
    status_code=status.HTTP_201_CREATED,
    tags=["videos"],
)
async def import_coco_pose(
    project_id: int,
    current_user: AdminUser,
    file: UploadFile = File(...),
    assignee_id: int | None = Form(None),
) -> dict:
    project = _require_project(project_id)
    if project.type.value != "pose_detection":
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "COCO-pose import is only available for pose_detection projects",
        )
    if assignee_id is not None:
        _require_user(assignee_id)
    try:
        return import_coco_service.import_coco_pose(
            project_id,
            file.file,
            file.filename or "import.zip",
            uploader_id=current_user.id,
            assignee_id=assignee_id,
        )
    except ValueError as e:
        code = (
            status.HTTP_413_REQUEST_ENTITY_TOO_LARGE
            if "larger than" in str(e).lower()
            else status.HTTP_400_BAD_REQUEST
        )
        raise HTTPException(code, str(e))
    except RuntimeError as e:
        raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, str(e))


@router.post(
    "/projects/{project_id}/import-images",
    status_code=status.HTTP_201_CREATED,
    tags=["videos"],
)
async def import_images(
    project_id: int,
    current_user: AdminUser,
    file: UploadFile = File(...),
    assignee_id: int | None = Form(None),
    resize_mode: str = Form("pad"),
) -> dict:
    project = _require_project(project_id)
    if project.type.value != "pose_detection":
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "Image import is only available for pose_detection projects",
        )
    if assignee_id is not None:
        _require_user(assignee_id)
    try:
        return image_import_service.import_images(
            project_id,
            file.file,
            file.filename or "images.zip",
            assignee_id=assignee_id,
            resize_mode=resize_mode,
        )
    except ValueError as e:
        code = (
            status.HTTP_413_REQUEST_ENTITY_TOO_LARGE
            if "larger than" in str(e).lower()
            else status.HTTP_400_BAD_REQUEST
        )
        raise HTTPException(code, str(e))
    except RuntimeError as e:
        raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, str(e))


@router.get("/projects/{project_id}/videos", tags=["videos"])
def list_videos(project_id: int, current_user: AdminUser) -> list[dict]:
    _require_project(project_id)
    return item_service.videos_in_project(project_id)


@router.patch(
    "/projects/{project_id}/videos/{source_video}/assign",
    tags=["videos"],
)
def reassign_video(
    project_id: int,
    source_video: str,
    data: ReassignRequest,
    current_user: AdminUser,
) -> dict:
    """Move frames of this video to another annotator, or back to the pool.

    Unfiltered this reassigns the whole video (the original behavior). With
    `from_assignee_id` / `only_unfinished` it becomes the "an annotator left"
    operation — see ReassignRequest.
    """
    _require_project(project_id)
    if data.assignee_id is not None:
        _require_user(data.assignee_id)
    if data.from_assignee_id is not None:
        # Validate rather than let a typo'd id quietly match no frame and
        # surface as "nothing to move".
        _require_user(data.from_assignee_id)
    moved, total = item_service.reassign_video(
        project_id,
        source_video,
        data.assignee_id,
        from_assignee_id=data.from_assignee_id,
        only_unfinished=data.only_unfinished,
    )
    if total == 0:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Video not found in project")
    if moved == 0:
        raise HTTPException(
            status.HTTP_409_CONFLICT,
            "No frame of this video matched the filters — nothing to reassign",
        )
    return {"reassigned": moved, "assignee_id": data.assignee_id}


@router.post(
    "/projects/{project_id}/videos/{source_video}/split",
    tags=["videos"],
)
def split_video(
    project_id: int,
    source_video: str,
    data: VideoSplitIn,
    current_user: AdminUser,
) -> dict:
    """Divide the video's unassigned frames into contiguous blocks, one per
    annotator. Frames that already have an owner are never moved."""
    _require_project(project_id)
    for assignee_id in data.assignee_ids:
        _require_user(assignee_id)
    result = item_service.split_video(project_id, source_video, data.assignee_ids)
    if result is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Video not found in project")
    if result["assigned"] == 0:
        raise HTTPException(
            status.HTTP_409_CONFLICT,
            "Every frame of this video already has an annotator — nothing to split",
        )
    return result


@router.post(
    "/projects/{project_id}/videos/{source_video}/rotate",
    tags=["videos"],
)
def rotate_video(
    project_id: int,
    source_video: str,
    data: RotateRequest,
    current_user: AdminUser,
) -> dict:
    _require_project(project_id)
    if data.degrees not in (90, 180, 270):
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "degrees must be 90, 180, or 270",
        )
    try:
        count = video_service.rotate_video(project_id, source_video, data.degrees)
    except RuntimeError as e:
        # ffmpeg failure — surface the message like upload_video does.
        raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, str(e))
    if count == 0:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Video not found in project")
    return {"rotated": count, "degrees": data.degrees}


@router.delete(
    "/projects/{project_id}/videos/{source_video}",
    tags=["videos"],
)
def delete_video(
    project_id: int,
    source_video: str,
    current_user: AdminUser,
) -> dict:
    _require_project(project_id)
    count = item_service.delete_video(project_id, source_video)
    if count == 0:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Video not found in project")
    return {"deleted": count}

File: frontend/src/api/videos.ts
import { api } from './client';

export interface VideoSummary {
  source_video: string;
  frames: number;
  done: number;       // annotated (done + reviewed combined)
  reviewed: number;   // strict subset that's been signed off
  unassigned: number; // frames still in the admin pool
  // null means unassigned OR mixed — after a split a video is mixed, so read
  // `unassigned` to tell the two apart.
  assigned_to: number | null;
}

export interface SplitBlock {
  assignee_id: number;
  count: number;
  first_frame: number | null;
  last_frame: number | null;
}

export interface SplitResult {
  assigned: number;
  skipped: number;
  blocks: SplitBlock[];
}

export type ResizeMode = 'stretch' | 'pad';

export async function uploadVideo(
  projectId: number,
  file: File,
  fps: number,
  assigneeId: number | null,
  rotation: 0 | 90 | 180 | 270 = 0,
  resizeMode: ResizeMode = 'pad',
) {
  const form = new FormData();
  form.append('file', file);
  form.append('fps', String(fps));
  if (assigneeId !== null) form.append('assignee_id', String(assigneeId));
  form.append('rotation', String(rotation));
  form.append('resize_mode', resizeMode);
  const { data } = await api.post<{
    video: string;
    frames: number;
    duration_s: number | null;
    expected_frames: number | null;
  }>(
    `/projects/${projectId}/videos`,
    form,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return data;
}

export async function listVideos(projectId: number) {
  const { data } = await api.get<VideoSummary[]>(`/projects/${projectId}/videos`);
  return data;
}

export async function reassignVideo(
  projectId: number,
  sourceVideo: string,
  assigneeId: number | null,
) {
  const { data } = await api.patch<{ reassigned: number; assignee_id: number | null }>(
    `/projects/${projectId}/videos/${encodeURIComponent(sourceVideo)}/assign`,
    { assignee_id: assigneeId },
  );
  return data;
}

/** Divide a video's unassigned frames into contiguous blocks, one per
 *  annotator, in the order given. Frames that already have an owner are
 *  left untouched. */
export async function splitVideo(
  projectId: number,
  sourceVideo: string,
  assigneeIds: number[],
) {
  const { data } = await api.post<SplitResult>(
    `/projects/${projectId}/videos/${encodeURIComponent(sourceVideo)}/split`,
    { assignee_ids: assigneeIds },
  );
  return data;
}

export interface CocoImportResult {
  items_created: number;
  annotations_created: number;
  skipped_images: number;
  skipped_labels: number;
  source_videos: string[];
}

export async function importCocoPose(
  projectId: number,
  file: File,
  assigneeId: number | null,
): Promise<CocoImportResult> {
  const form = new FormData();
  form.append('file', file);
  if (assigneeId !== null) form.append('assignee_id', String(assigneeId));
  const { data } = await api.post<CocoImportResult>(
    `/projects/${projectId}/import-coco`,
    form,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return data;
}

export interface ImageImportResult {
  items_created: number;
  skipped_files: number;
  source_video: string;
  resize_mode: ResizeMode;
}

export async function importImages(
  projectId: number,
  file: File,
  assigneeId: number | null,
  resizeMode: ResizeMode = 'pad',
): Promise<ImageImportResult> {
  const form = new FormData();
  form.append('file', file);
  if (assigneeId !== null) form.append('assignee_id', String(assigneeId));
  form.append('resize_mode', resizeMode);
  const { data } = await api.post<ImageImportResult>(
    `/projects/${projectId}/import-images`,
    form,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return data;
}

export async function deleteVideo(projectId: number, sourceVideo: string) {
  const { data } = await api.delete<{ deleted: number }>(
    `/projects/${projectId}/videos/${encodeURIComponent(sourceVideo)}`,
  );
  return data;
}

export async function rotateVideo(
  projectId: number,
  sourceVideo: string,
  degrees: 90 | 180 | 270,
) {
  const { data } = await api.post<{ rotated: number; degrees: number }>(
    `/projects/${projectId}/videos/${encodeURIComponent(sourceVideo)}/rotate`,
    { degrees },
  );
  return data;
}
